### ISSUE TITLE
Policy Gradient Understanding

### ISSUE THREAD

**Issue by yufengm** [ruotianluo/self-critical.pytorch#18]
Hi Ruotian,

Thanks for implementing the self critical sequence training in pytorch. It's really helpful.

I just have some concerns about the policy gradient computation here. In your code
https://github.com/ruotianluo/self-critical.pytorch/blob/fd445dc935e1a474150107faa4c36be81415892d/misc/utils.py#L50, it seems that you are just averaging over the log probabilities of the all words. While from my understanding and in the paper, this seems to be the average of each review's log probabilities. Is there anything wrong with my understanding here?

Thanks for your time.
Yufeng

**Comment by ruotianluo** [comment]
It's the averaging of reward times log probability. Check out the previous line.

**Comment by yufengm** [comment]
Yeah. But it seems that you were dividing by the total number of generated words(number of 1s in mask), right? What I am thinking was to divide by the number of captions, i.e., the batch size, as we want the expectation over reward for a whole caption. Am I misunderstanding here?

Thanks for your time. I really appreciate it.

**Comment by ruotianluo** [comment]
This is not related to reinforce actually. The cross entropy loss I use here is also not "standard" cross entropy loss. All the losses I use here are word-wise instead of caption-wise. This is inherited from original neuraltalk2 reprository; my interpratation of this is the caption-wise loss will focus more on shorter captions.
In fact, if you divide it by the number of captions, the result may not be very different.

**Comment by yufengm** [comment]
Alright, thanks. Yeah word-wise cross entropy would be more accurate in terms of caption length.

**Comment by Yugnaynehc** [comment]
I have the same question and I think @yufengm you are right.
